selfservice/strategy/code/strategy_recovery_test.go

```diff
@@ -0,0 +1,1003 @@
+		actual := expectSuccessfulRecovery(t, cl, RecoveryFlowTypeBrowser, values)
+		message := testhelpers.CourierExpectMessage(t, reg, email, "Recover access to your account")
+		recoveryCode := testhelpers.CourierExpectCodeInMessage(t, message, 1)
+
+		cl.CheckRedirect = func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		}
+
+		action := gjson.Get(actual, "ui.action").String()
+		require.NotEmpty(t, action)
+		csrf_token := gjson.Get(actual, "ui.nodes.#(attributes.name==csrf_token).attributes.value").String()
+		require.NotEmpty(t, csrf_token)
+
+		res, err := cl.PostForm(action, url.Values{
+			"code":       {recoveryCode},
+			"csrf_token": {csrf_token},
+		})
+		require.NoError(t, err)
+		require.NoError(t, res.Body.Close())
+		assert.Equal(t, http.StatusSeeOther, res.StatusCode)
+		require.Len(t, cl.Jar.Cookies(urlx.ParseOrPanic(public.URL)), 2)
+		cookies := spew.Sdump(cl.Jar.Cookies(urlx.ParseOrPanic(public.URL)))
+		assert.Contains(t, cookies, "ory_kratos_session")
+
+		actualSession, err = reg.SessionPersister().GetSession(context.Background(), sess.ID)
+		require.NoError(t, err)
+		assert.False(t, actualSession.IsActive())
+	})
+
+	t.Run("description=should not be able to use an invalid code more than 5 times", func(t *testing.T) {
+		c := testhelpers.NewClientWithCookies(t)
+		f := testhelpers.InitializeRecoveryFlowViaBrowser(t, c, false, public, nil)
+
+		form := testhelpers.SDKFormFieldsToURLValues(f.Ui.Nodes)
+		form.Set("code", "123123")
+
+		for submitTry := 0; submitTry < 5; submitTry++ {
+			res, err := c.PostForm(f.Ui.Action, form)
+			require.NoError(t, err, "try=%d", submitTry)
+			assert.Equal(t, http.StatusOK, res.StatusCode, "try=%d", submitTry)
+			assert.Contains(t, res.Request.URL.String(), conf.SelfServiceFlowRecoveryUI(ctx).String()+"?flow=")
+
+			rs, _, err := testhelpers.
+				NewSDKCustomClient(public, c).
+				V0alpha2Api.GetSelfServiceRecoveryFlow(context.Background()).
+				Id(res.Request.URL.Query().Get("flow")).Execute()
+
+			require.NoError(t, err, "try=%d", submitTry)
+
+			require.Len(t, rs.Ui.Messages, 1, "try=%d", submitTry)
+			assert.Equal(t, "The recovery code is invalid or has already been used. Please try again.", rs.Ui.Messages[0].Text, "try=%d", submitTry)
+		}
+
+		// submit an invalid code for the 6th time
+		res, err := c.PostForm(f.Ui.Action, form)
+		require.NoError(t, err)
+		assert.Equal(t, http.StatusOK, res.StatusCode)
+
+		rs, _, err := testhelpers.
+			NewSDKCustomClient(public, c).
+			V0alpha2Api.GetSelfServiceRecoveryFlow(context.Background()).
+			Id(res.Request.URL.Query().Get("flow")).Execute()
+
+		require.NoError(t, err)
+
+		require.Len(t, rs.Ui.Messages, 1)
+		assert.Equal(t, "The recovery was submitted too often. Please try again.", rs.Ui.Messages[0].Text)
+		json, err := rs.Ui.MarshalJSON()
+		require.NoError(t, err)
+		t.Logf("body: %s", string(json))
+		assert.True(t, gjson.GetBytes(json, "nodes.#(attributes.name==email)").Exists())
```

Comment:
Add an assertion that we get a fresh flow :)

Reply:
Done.